### ISSUE TITLE
Automatically translate parameter reference to expr?

### ISSUE THREAD

**Issue by jheer** [vega/vega-lite#7300]
There are different ways to refer to parameters in Vega-Lite v5:
1. For predicates and data extents, we support `{ "param": "paramName" }`
2. Elsewhere, we support arbitrary expressions such as `{ "expr": "'paramName'" }`

Given this discrepancy, it might be more usable and less error-prone to automatically translate parameter references (option 1 above) to simple expressions (option 2) when parameter references are provided within a context that expects expressions.

Doing so would make it much easier to support symbolic reference to parameters in APIs (Vega-Lite JS API, Altair, etc), as those APIs would only need to handle two cases (parameter definition and parameter reference) and would not need to include internal tracking of reference versus expression contexts. In addition, this change could help prevent issues such as #7284, which I expect will recur.

So the main question is if this is feasible within the VL internals. Is there a clean separation between expression contexts and parameter reference contexts? If so, I'd love to have `{ param: "paramName" }` automatically translated to `{ expr: "'paramName'" }` within expression contexts.

**Comment by domoritz** [comment]
Just checking, do you mean translate `{ "param": "paramName" }` to `{ "expr": "paramName" }` and rewriting selection params to their respective predicate check?

@arvind @kanitw and I discussed this today and think that we could rewrite selection param references to their predicate checks in arbitrary expressions and use `expr` everywhere (and deprecate param). This change should make the experience more consistent. What do you think?

**Comment by jheer** [comment]
Yes, I mean translate `{ "param": "paramName" }` to `{ "expr": "paramName" }`, _in those contexts where expressions are supported but `{ "param": "paramName" }` is not_.

Deprecating `param` in favor of `expr` would solve the API issue I'm encountering. I'm not sure if it helps or hurts usability overall, in the sense that `param` has a more obvious connection to the `params` definitions than `expr`. But that is a question for VL proper, as it won't really be an issue in the API (both `vl.param()` and `vl.expr()` references will work there).
